src/rpc/rawtransaction.cpp

```diff
@@ -922,67 +924,77 @@ static RPCHelpMan testmempoolaccept()
+            return AcceptToMemoryPool(mempool, txns[0], false /* bypass_limits */, true /* test_accept */)));
+    } else {
+        validation_results = WITH_LOCK(cs_main, return ProcessNewPackage(mempool, txns, true));
     }
-    result_0.pushKV("allowed", test_accept_res);
 
-    // Only return the fee and vsize if the transaction would pass ATMP.
-    // These can be used to calculate the feerate.
-    if (test_accept_res) {
-        result_0.pushKV("vsize", virtual_size);
-        UniValue fees(UniValue::VOBJ);
-        fees.pushKV("base", ValueFromAmount(fee));
-        result_0.pushKV("fees", fees);
-    } else {
-        if (state.IsInvalid()) {
-            if (state.GetResult() == TxValidationResult::TX_MISSING_INPUTS) {
-                result_0.pushKV("reject-reason", "missing-inputs");
+    auto tx_accepted = [](MempoolAcceptResult& res) { return res.m_accepted; };
+    const bool all_valid = std::all_of(validation_results.begin(), validation_results.end(), tx_accepted);
+    // ProcessNewPackage should return a MempoolAcceptResult per transaction
+    // or exactly 1 for the first error that occurs.
+    CHECK_NONFATAL(all_valid || validation_results.size() == 1);
+
+    // TODO: Report absurd fees for packages.
+
+    UniValue result(UniValue::VARR);
+    if (all_valid) {
+        for (unsigned int i = 0; i < validation_results.size(); ++i) {
+            UniValue result_inner(UniValue::VOBJ);
+            const CTransaction tx = *txns[i];
+            result_inner.pushKV("txid", tx.GetHash().GetHex());
+            result_inner.pushKV("wtxid", tx.GetWitnessHash().GetHex());
+
+            const CAmount fee = validation_results[i].m_base_fees.value();
+            const int64_t virtual_size = GetVirtualTransactionSize(tx);
+            const CAmount max_raw_tx_fee = max_raw_tx_fee_rate.GetFee(virtual_size);
+            // Check that fee does not exceed maximum fee
+            if (max_raw_tx_fee && fee > max_raw_tx_fee) {
+                result_inner.pushKV("allowed", false);
+                result_inner.pushKV("reject-reason", "max-fee-exceeded");
+                result_inner.push_back(std::move(result_inner));
             } else {
-                result_0.pushKV("reject-reason", strprintf("%s", state.GetRejectReason()));
+                result_inner.pushKV("allowed", true);
+                result_inner.pushKV("vsize", virtual_size);
+                UniValue fees(UniValue::VOBJ);
+                fees.pushKV("base", ValueFromAmount(fee));
+                result_inner.pushKV("fees", fees);
             }
+            result.push_back(std::move(result_inner));
+        }
+    } else {
+        Assume(validation_results[0].m_failed_ptx);
+        UniValue result_0(UniValue::VOBJ);
+        result_0.pushKV("txid", validation_results[0].m_failed_ptx->GetHash().GetHex());
+        result_0.pushKV("wtxid", validation_results[0].m_failed_ptx->GetWitnessHash().GetHex());
```

Comment:
I'm not sure that I like the interface of passing back a CTransactionRef just for this failure case. I wonder if we should:

- allow `m_accepted` to be tri-state {accepted, failed, not_fully_validated}. That can be achieved with a `std::optional<bool>`
- return a `MempoolAcceptResult` for all transactions in the package where either:
  - all are success
  - one is fail and the rest are not_fully_validated
- work out the txid of the failing transaction from its index in the vector.

What do you think?